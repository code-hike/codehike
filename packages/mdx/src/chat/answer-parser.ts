// turns streaming markdown into codeblocks, content and replies
// doesnt care about previous answers
export function parseAnswer(
  markdown: string,
  isStreaming: boolean
) {
  const { markdownWithoutCode, fileInfoList } =
    extractCodeBlocks(markdown, isStreaming)
  const [answerText, repliesText] =
    markdownWithoutCode.split(/\n+---\n+/)
  const replies = repliesText
    ? repliesText
        .split(/\n/)
        .map((r: string) => r.replace(/^-\s*/, "").trim())
        .filter((r: any) => r !== "")
    : []

  return {
    fileInfoList: fileInfoList,
    content: answerText.replace(/\n-+$/, ""),
    replies,
  }
}

function extractCodeBlocks(
  markdown: string,
  isStreaming: boolean
) {
  const closedCodeBlocks =
    markdown.match(/```[\s\S]*?```/g) || []
  const markdownWithoutClosedCodeBlocs = markdown.replace(
    /```[\s\S]*?```/g,
    ""
  )

  const fileInfoList = closedCodeBlocks.map(s => ({
    ...codeblockToFileInfo(s),
    open: false,
  }))

  let markdownWithoutCode = markdownWithoutClosedCodeBlocs

  if (isStreaming) {
    const openCodeBlock =
      markdownWithoutClosedCodeBlocs.match(/```[\s\S]*?$/g)

    if (openCodeBlock) {
      markdownWithoutCode =
        markdownWithoutClosedCodeBlocs.replace(
          /```[\s\S]*?$/g,
          ""
        )

      let streamingCodeBlock = openCodeBlock[0]
        .split("\n")
        .slice(0, -1)
        .join("\n")

      if (streamingCodeBlock) {
        fileInfoList.push({
          ...codeblockToFileInfo(
            streamingCodeBlock + "\n```"
          ),
          open: isStreaming,
        })
      }
    }
  }

  // if ends with ` or ``, let's wait until more chars
  const backticks = markdownWithoutCode.match(/(`$|``$)/g)
  if (backticks) {
    markdownWithoutCode = markdownWithoutCode.replace(
      /(`$|``$)/g,
      ""
    )
  }

  markdownWithoutCode = markdownWithoutCode.trim()
  // console.log({ markdownWithoutCode })

  return { markdownWithoutCode, fileInfoList }
}

function codeblockToFileInfo(codeblock: string): {
  lang: string
  name: string
  text: string
} {
  const codeBlockPattern =
    /```([^ \n]+)? ?([^\n]+)?\n([\s\S]*?)\n?```/g

  const match = codeBlockPattern.exec(codeblock)

  return {
    lang: match?.[1] || "",
    name: match?.[2] || "answer",
    text: match?.[3] || "",
  }
}
