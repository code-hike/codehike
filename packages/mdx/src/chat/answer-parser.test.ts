import { expect, test } from "vitest"
import { parseAnswer } from "./answer-parser"

const fixtures = [
  ``,
  `Hello`,
  `
~~~js foo.js
console.log(1)
~~~

# hello

world

---

- yes
- no

`,
  `
~~~js foo.js
console.log(1)

~~~

~~~ts bar.ts
console.log(2)
~~~

# hello

world`,
  `
~~~js foo.js
console.log(1)

`,
  `~~~js foo.js
console.log(1)
~~~

Hello

-`,
  `~~~js foo.js
console.log(1)
~~~

Hello

---`,
  ,
  `~~~js foo.js
console.log(1)
~~~

Hello

---
`,
].map(x => x.replace(/~/g, "`"))

test.each(fixtures)("%s", async markdown => {
  expect(parseAnswer(markdown, true)).toMatchSnapshot()
})
