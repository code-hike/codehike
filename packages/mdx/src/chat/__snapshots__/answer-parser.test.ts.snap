// Vitest Snapshot v1

exports[`
\`\`\`js foo.js
console.log(1)

 1`] = `
{
  "content": "",
  "fileInfoList": [
    {
      "lang": "js",
      "name": "foo.js",
      "open": true,
      "text": "console.log(1)
",
    },
  ],
  "replies": [],
}
`;

exports[`
\`\`\`js foo.js
console.log(1)

\`\`\`

\`\`\`ts bar.ts
console.log(2)
\`\`\`

# hello

world 1`] = `
{
  "content": "# hello

world",
  "fileInfoList": [
    {
      "lang": "js",
      "name": "foo.js",
      "open": false,
      "text": "console.log(1)
",
    },
    {
      "lang": "ts",
      "name": "bar.ts",
      "open": false,
      "text": "console.log(2)",
    },
  ],
  "replies": [],
}
`;

exports[`
\`\`\`js foo.js
console.log(1)
\`\`\`

# hello

world

---

- yes
- no

 1`] = `
{
  "content": "# hello

world",
  "fileInfoList": [
    {
      "lang": "js",
      "name": "foo.js",
      "open": false,
      "text": "console.log(1)",
    },
  ],
  "replies": [
    "yes",
    "no",
  ],
}
`;

exports[` 1`] = `
{
  "content": "",
  "fileInfoList": [],
  "replies": [],
}
`;

exports[`\`\`\`js foo.js
console.log(1)
\`\`\`

Hello

- 1`] = `
{
  "content": "Hello
",
  "fileInfoList": [
    {
      "lang": "js",
      "name": "foo.js",
      "open": false,
      "text": "console.log(1)",
    },
  ],
  "replies": [],
}
`;

exports[`\`\`\`js foo.js
console.log(1)
\`\`\`

Hello

---
 1`] = `
{
  "content": "Hello
",
  "fileInfoList": [
    {
      "lang": "js",
      "name": "foo.js",
      "open": false,
      "text": "console.log(1)",
    },
  ],
  "replies": [],
}
`;

exports[`\`\`\`js foo.js
console.log(1)
\`\`\`

Hello

--- 1`] = `
{
  "content": "Hello
",
  "fileInfoList": [
    {
      "lang": "js",
      "name": "foo.js",
      "open": false,
      "text": "console.log(1)",
    },
  ],
  "replies": [],
}
`;

exports[`Hello 1`] = `
{
  "content": "Hello",
  "fileInfoList": [],
  "replies": [],
}
`;
